### DISCUSSION TITLE
VipsForeignSave: ".gif" is not a known buffer format

### DISCUSSION THREAD

**Discussion by AndKe** [libvips/libvips#2189] [Q&A]
The  buf = x.write_to_buffer('.png')  works fine for me.
In attempt to make "lighter" images for faster processing(unpacking with limited CPU power), I tried to make .gif 

GIF's limited color space could maybe be faster..
I was under the impression that gif's are supported?

All I get is this:
```
  File "/home/andre/prog/PycharmProjects/PFD/PFD-perf.py", line 640, in download
    buf = x.write_to_buffer('.gif')
  File "/home/andre/.local/lib/python3.8/site-packages/pyvips/vimage.py", line 591, in write_to_buffer
    raise Error('unable to write to buffer')
pyvips.error.Error: unable to write to buffer
  VipsForeignSave: ".gif" is not a known buffer format
```

**Comment by jcupitt** [comment]
Hi @AndKe,

GIF load is built into libvips, but it delegates GIF save to imagemagick, and imagemagick needs to be told the save format. Try:

```
$ python3
Python 3.8.6 (default, Jan 27 2021, 15:42:20) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("k2.jpg")
>>> y = x.write_to_buffer(".gif", format="GIF")
>>> 
```

**Reply by AndKe** [reply]
@jcupitt  thank you, unfortunately, this did not work:
```
  File "/home/andre/prog/PycharmProjects/PFD/PFD-perf.py", line 642, in download
    buf = x.write_to_buffer(".gif", format="GIF")
  File "/home/andre/.local/lib/python3.8/site-packages/pyvips/vimage.py", line 591, in write_to_buffer
    raise Error('unable to write to buffer')
pyvips.error.Error: unable to write to buffer
  VipsForeignSave: ".gif" is not a known buffer format
```
Is there something else I should try ?

**Comment by jcupitt** [comment]
Perhaps your libvips has been built without imagemagic support? Try this at the command-line:

```
$ vips magicksave
save file with ImageMagick
usage:
   magicksave in filename [--option-name option-value ...]
where:
   in           - Image to save, input VipsImage
   filename     - Filename to save to, input gchararray
optional arguments:
   format       - Format to save in, input gchararray
...
```

**Reply by AndKe** [reply]
I get:
```
$ vips magicsave
vips: unknown action "magicsave"
```

My vips is installed using pip install pyvips, version 2.1.14 on Ubuntu 20.10  - can it really be missing some essential parts?

**Comment by jcupitt** [comment]
You need a "k" in there. Try copy-pasting the command.

**Comment by AndKe** [comment]
@jcupitt  I am sorry (for the wrong copy-n-paste), this was early in the morning, and before driving to work, I did try it correctly, before trying what you saw (in case it was misspelled by accident)

```
$ vips magicksave
vips: unknown action "magicksave"

```
do you have any other suggestions ?

**Comment by jcupitt** [comment]
That means your libvips has been built without imagemagick support, so you won't be able to save GIFs.

Which vips are you using? The one that comes with ubuntu 20.10 does have imagemagick:

```
john@banana ~ $ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.10
Release:	20.10
Codename:	groovy
john@banana ~ $ which vips
/usr/bin/vips
john@banana ~ $ vips magicksave
save file with ImageMagick
usage:
   magicksave in filename [--option-name option-value ...]
where:
   in           - Image to save, input VipsImage
   filename     - Filename to save to, input gchararray
optional arguments:
   format       - Format to save in, input gchararray
...
```

**Comment by AndKe** [comment]
you are right, a few days ago I was led to install from git,  :)  
So mine was in /usr/local/bin/ , not the version that is in "Groovy".

I have removed all vips files from /usr/local/bin/  and /usr/local/lib/  and /usr/local/include/vips/
and installed using apt  - and it works.
Thank you for all the help, and keep up the good work, I just "discovered" pyvips via another project, and realize it's power.

**Comment by AndKe** [comment]
@jcupitt  One more question:
Now that I got GIF to work fine (previously used PNG)  - I see a performance problem compared to .png.
The process is generating 640×360 <16 color svg which is converted to PNG. (it runs on RPi4)
The speed drops significantly if converting to GIF instead . - yet , the recepient (Google Glass browser) decode the GIF much faster/with less CPU.
It seems to me that the GIF compression is less optimal than PNG.  Is there anything you can think of that can improve the svg->gif performance?  - or is PNG8 supported? -or any other suitable format you can think of?

**Comment by jcupitt** [comment]
Yes, gif is slow to compress.

You could try uncompressed PNG -- it's quite a bit quicker. Just set `compression=0` to PNG write. Your files will be a bit larger, of course.

```
$ time vips copy k2.jpg x.png[compression=0]

real	0m0.152s
user	0m0.153s
sys	0m0.027s
$ time vips copy k2.jpg x2.png

real	0m1.005s
user	0m0.986s
sys	0m0.040s
$ ls -l x.png x2.png
-rw-r--r-- 1 john john 3255372 Apr  9 10:19 x2.png
-rw-r--r-- 1 john john 8925331 Apr  9 10:19 x.png
```

Yes, libvips supports png8, but it's pretty slow.

**Comment by AndKe** [comment]
Thank you.
May I ask one more thing (I wish to make a synthetic test that is as close to the real compression job as possible)
The code that do call visa looks like:
```
    def download(self, update_index=0):
        self.attributes['xmlns']="http://www.w3.org/2000/svg"
        content = gui.Widget.repr(self, None)
        #headers = {'Content-type': 'image/svg+xml'}  
        headers = {'Content-type': 'image/png'}
        x = pyvips.Image.svgload_buffer(content.encode('utf8'))
        buf = x.write_to_buffer('.png')
        #buf = x.write_to_file('a.png')
```
I believe the source is a svg (in memory) and normally written to buffer as png
You can see (last line) I played with saving to png.
The question is: how can I do a write_to_file and save it as the original data? (so that I can use that file for synthetic benchmarks)

**Reply by jcupitt** [reply]
Do you mean save the SVG that was downloaded as an SVG?

**Reply by AndKe** [reply]
Yes, I think you understood me correctly -  there is a svg being generated by this python app, and (not "downloaded" in the sense I think of it)  
To me , it seems pyvips reads it here: x = pyvips.Image.svgload_buffer(content.encode('utf8'))
and stores it using  buf = x.write_to_buffer('.png')  (or maybe I am wrong about the first one?)
I would like to save it (the svg) exactly as vips sees it before compression - so that I could time various output formats and compressions/quality using "time vips copy..."

**Reply by jcupitt** [reply]
You should just be able to do something like:

```python
with open("/tmp/test.svg", "w") as f:
    f.write(content)
x = pyvips.Image.svgload_buffer(content.encode('utf8'))
...
```

**Reply by AndKe** [reply]
Once again: Thank you! :)
